fe/fe-core/src/main/java/com/starrocks/binlog/BinlogManager.java

```diff
@@ -0,0 +1,245 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.binlog;
+
+import com.starrocks.alter.SchemaChangeHandler;
+import com.starrocks.catalog.Database;
+import com.starrocks.catalog.OlapTable;
+import com.starrocks.catalog.Partition;
+import com.starrocks.catalog.Table;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.common.DdlException;
+import com.starrocks.common.util.PropertyAnalyzer;
+import com.starrocks.persist.ModifyTablePropertyOperationLog;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.thrift.TTabletMetaType;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class BinlogManager {
+
+    private static final Logger LOG = LogManager.getLogger(BinlogManager.class);
+
+    // statics tablet through report handler
+    // tableId -> tabletId -> beId
+    private Map<Long, Map<Long, Set<Long>>> tabletStatics;
+
+    // tableId -> number of reported tablet from different BE
+    private Map<Long, Long> tableIdToReportedNum;
+
+    // tableId -> binlog version
+    private Map<Long, Long> tableIdToBinlogVersion;
+
+    private ReentrantReadWriteLock binlogLock = new ReentrantReadWriteLock(true);
+
+    public static final long INVALID = -1;
+
+    public BinlogManager() {
+        tableIdToReportedNum = new HashMap<>();
+        tableIdToBinlogVersion = new HashMap<>();
+        tabletStatics = new HashMap<>();
+    }
+
+    // the caller hold the write lock of Database
+    public void checkAndSetBinlogAvailableVersion(Database db, OlapTable table, long tabletId,
+                                                  long beId, long totalNum) {
+        tabletStatics.putIfAbsent(table.getId(), new HashMap<>());
+        Map<Long, Set<Long>> allTabletIds = tabletStatics.get(table.getId());
+
+        // the previous version binlog config was not completed,
+        // and a new version was generated,
+        // the tablets counted need to be cleared
+        if (tableIdToBinlogVersion.containsKey(tabletId) &&
+                tableIdToBinlogVersion.get(tabletId) != table.getBinlogVersion()) {
+            allTabletIds.clear();
+            tableIdToReportedNum.remove(table.getId());
+        }
+        tableIdToBinlogVersion.put(table.getId(), table.getBinlogVersion());
+        tableIdToReportedNum.putIfAbsent(table.getId(), 0L);
+        long num = tableIdToReportedNum.get(table.getId());
+
+        allTabletIds.putIfAbsent(tabletId, new HashSet<>());
+        Set<Long> allBeIds = allTabletIds.get(tabletId);
```

Comment:
`putIfAbsent` will return the set, and no need to get it again

Reply:
if the tablet exits already，`putIfAbsent` will return null。

Reply:
use `computIfAbsent`?